src/sync/bounded.zig

```diff
@@ -0,0 +1,995 @@
+                    if (self.tryAcquireReceiveSlot(&temp_slot)) {
+                        return self.read(&temp_slot);
+                    } else |err| switch (err) {
+                        // if channel disconnected, we return as nothing else can be done
+                        error.disconnected => return error.disconnected,
+                        // if channel empty, we backoff before putting thread to sleep
+                        error.empty => {
+                            if (backoff.isCompleted()) {
+                                break;
+                            } else {
+                                backoff.snooze();
+                            }
+                        },
+                    }
+                }
+
+                if (timeout != null and (std.time.Instant.now() catch unreachable).order(timeout.?) == .gt)
+                    return error.timeout;
+
+                var thread_ctx = thread_context.getThreadLocalContext();
+                thread_ctx.reset();
+                const opId = temp_slot.toOperationId();
+                self.receivers.registerOperation(opId, thread_ctx);
+
+                // We do this because if channel is not full or if channel is not disconnected,
+                // (in either case) we don't want to wait. Let's break out of the sleep.
+                //
+                // If some other operation was able to update this context's state, we are ok
+                // with this. This is why we don't check the return value as a call to `waitUntil`
+                // will allow us to perform the checks below.
+                if (!self.isEmpty() or self.isDisconnected())
+                    _ = thread_ctx.tryUpdateFromWaitingStateTo(.aborted);
+
+                switch (thread_ctx.waitUntil(timeout)) {
+                    .waiting => unreachable,
+                    .aborted, .disconnected => {
+                        // there must be an entry with this operation id, if not panic!
+                        _ = self.receivers.unregisterOperation(opId) orelse unreachable;
+                    },
+                    .operation => |operationId| {
+                        std.debug.assert(opId == operationId);
+                    },
+                }
+            }
+        }
+
+        /// Acquires a receiver modifying the channel state.
+        pub inline fn acquireReceiver(self: *Self) void {
+            _ = self.n_receivers.fetchAdd(1, .seq_cst);
+        }
+
+        /// Releases a receiver modifying the channel state. Returns `true` if the
+        /// receiver was successfully released else returns `false` if `n_recievers < 1`
+        /// indicating a (potential) invalid state.
+        pub inline fn releaseReceiver(self: *Self) bool {
+            if (self.n_receivers.load(.seq_cst) == 0)
+                return false;
+            if (self.n_receivers.fetchSub(1, .seq_cst) == 1) {
+                self.disconnect();
+            }
```

Comment:
nit: can we be consistent with if statement brackets here

ie, theres no brackets here, but two lines down theres brackets

on a more general note: idk if its worth a discussion but i would vote we *always* use brackets in if statements across the entire codebase - imo the option to bracket or not bracket single line blocks goes against the simplicity/'one way to do it' of zig which i do not like

Reply:
In this specific code example, I agree that it's best to use braces for both. Typically, I prefer to use braces in `if` *statements*.

In `if` *expressions* (when an `if-else` evaluates to a non-void value), I would argue to only use braces if it is actually required. For example:
```zig
const max = if (a > b) a else b;
```
Using braces, it would need to be be written like this:
```zig
const max = if (a > b) blk: {
    break :blk a;
} else blk: {
    break :blk b;
};
```
Braces would only be required if you need to execute multiple statements for each condition, since multiple statements would need to be executed in a block.